_data-prepper/common-use-cases/event-aggregation.md

```diff
@@ -0,0 +1,135 @@
+---
+layout: default
+title: Event aggregation with Data Prepper
+parent: Common use cases
+nav_order: 40
+---
+
+# Event aggregation with Data Prepper
+
+You can use Data Prepper to aggregate data from different events over a period of time. Aggregating events can help reduce unnecessary log volume and handle use cases like multiline logs that come in as separate events. The [Aggregate processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/aggregate/) is a stateful processor that groups events based on the values for a set of specified identification keys, and performs a configurable action on each group.
+
+State in the Aggregate processor is stored in memory. For example, in order to combine four events into one, the processor needs to retain pieces of the first three events. The state of an aggregate group of events is kept for a configurable amount of time. Depending on your logs, the aggregate action being used, and the amount of memory options in the processor configuration, the aggregation could take place over a long period of time.
+
+## Basic usage
+
+The following example pipeline extracts the fields `sourceIp`, `destinationIp`, and `port` using the [Grok processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/grok/), and then aggregates on those fields over a period of 30 seconds using the [Aggregate processor]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/configuration/processors/aggregate/) and the `put_all` action. At the end of the 30 seconds, the aggregated log is sent to the OpenSearch sink.
+
+```json
+aggregate_pipeline:  
+   source:
+     http:
+      path: "/${pipelineName}/logs"
+   processor:
+     - grok:
+         match: 
+           log: ["%{IPORHOST:sourceIp} %{IPORHOST:destinationIp} %{NUMBER:port:int}"]
+     - aggregate:
+         group_duration: "30s"
+         identification_keys: ["sourceIp", "destinationIp", "port"]
+         action:
+           put_all:
+   sink:
+     - opensearch:
+         ...
+         index: aggregated_logs
+```
+{% include copy-curl.html %}
+
+For example, consider the following batch of logs:
+
+```json
+{ "log": "127.0.0.1 192.168.0.1 80", "status": 200 }
+{ "log": "127.0.0.1 192.168.0.1 80", "bytes": 1000 }
+{ "log": "127.0.0.1 192.168.0.1 80" "http_verb": "GET" }
+```
+{% include copy-curl.html %}
+
+The Grok processor will extract the `identification_keys` to create the following logs:
```

Comment:
The grok processor is not creating the following logs. It is modifying the existing log events to look like what follows. Also, it is extract keys more generally. The pipeline author has to extract them to match what `aggregate` expects for `identification_keys`.

```
The grok processor will extract extract keys such that the log events look like the following. These events now have the data which the aggregate processor will need for the `identification_keys`.
```